pilot/pkg/model/sidecar.go

```diff
@@ -217,7 +217,27 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config, configNamespa
 	// Assign namespace dependencies
 	out.namespaceDependencies = make(map[string]struct{})
 	for _, listener := range out.EgressListeners {
-		for _, s := range listener.services {
+		// Explicitly callable services are potential destinations, copy those first
+		serviceDestinations := append([]*Service(nil), listener.services...)
+
+		// Infer more possible destinations from virtual services
+		for _, vs := range listener.virtualServices {
+			v := vs.Spec.(*networking.VirtualService)
+			for _, d := range virtualServiceDestinations(v) {
+				// Default to this hostname in our config namespace
+				if s, ok := ps.ServiceByHostnameAndNamespace[host.Name(d.Host)][configNamespace]; ok {
+					serviceDestinations = append(serviceDestinations, s)
+				} else {
+					// Otherwise pick randomly
```

Comment:
I think this may not be 100% accurate. Some background here, you can have the same hostname in different namespaces with ServiceEntries. So for example you may have google.com defined in ns A and ns B, potentially with different configs (say team A only sends http traffic and team B only sends https traffic so they only define ports 80/443 as an example).

The design here was that if you explicitly select one in your sidecar egress (eg `a/*` or `a/google.com`), you should get that one. So we would actually want to iterate over `listener.listenerHosts` potentionally? 

However looking closer, this seems like the wrong place to do so. Should we move all of this logic into the `selectServices` method that builds up listener.services? I think that may make more sense since its already doing all of this logic I mentioned above

Reply:
I had thought of this too. If we are auto importing the services, then when we hit this point where the service is not in local namespace, we need to decide what to do when the service exists in multiple namespaces. I think in this case, we should not import anything as we can't make some decision arbitrary on the user's behalf. What we should do here is check if there is ONLY ONE service outside the config namespace. if there is only one, then we import it and if not, we stop importing it. 

Of course, it does mean that the moment someone adds a second service entry for google.com in other namespaces, existing *automatic* imports would break :(. 

I think what we should do is the following:
 1. if the service is present in the local `configNamespace` pick it.
 2. if it is not in the local config namespace, we look for the service in the importing virtual service's namespace, and import it if present. If the service is not present in the importing virtual service's namespace, then I think we should punt and not import the service from some other arbitrary namespace, due to the issue I mentioned above. The only exception to the rule is Kubernetes service (check the service attribute) - but I think we can punt this piece for now, if @jackkleeman 's use case is to simply import the services from the virtual service's namespace.

Reply:
I agree with you @rshriram. In our use case, I think the sidecar will be in the pod namespace, the virtualservice will be in `istio-config`, and the services are kube services and can be anywhere. we'll definitely have some cross namespace traffic. so I think I will have to special case k8s services for that to work for me

Reply:
can we not figure out here what destination will actually get used by the virtualservice?

Reply:
@rshriram we discussed this for months, lets not come up with a new resolution strategy. What I described is not just how I think it should work, it is how the code is implemented today. We need to follow how the code is implemented today or we will end up selecting destinations that are not used by the virtualservice.

Reply:
ie, pick configNamespace, otherwise pick randomly

Reply:
hmm, actually I think I might see what @howardjohn is saying. we need to prioritise picking a host from the existing `listener.services` field, in case someone has specifically asked for that hostname

Reply:
oh actually, I think we already effectively have that; we only add a service if the hostname doesn't exist in `servicesAdded`, and we process the list of specifically asked for services first. so if you have `eg a/* or a/google.com`, you'll get that host, because it will win the race to `servicesAdded`

Reply:
To clarify:

The order is:
* In config namespace
* Imported in egress (if multiple, random)
* random

Based on what you said, I think you are right we do respect this. Can we just add a unit test/comment since this is pretty subtly and likely to break if we change in the future

Reply:
@howardjohn I am not changing the model. My proposal actually aligns with the current model and doesn't introduce a new one. Here is an example:

```
 istio-config has a vs for foo.com that refers to s1.ns1 and s2.ns2, and google.com as destinations
 google.com is declared as a service entry in ns1 and ns2 namespaces
 the Sidecar in ns3 namespace imports from istio-config/foo.com
```

- we first process the sc.Services() without this new logic. This would take care of importing all the hosts with all the conflict resolutions as necessary.
- Now we process all the virtual services. For each destination, we check if that destination is already present in the existing set (sc.Services()). If it is, then great. If not, we follow the logic below:

- we look for s1.ns1 in pod's ns3 namespace (none found), then look for it in istio-config namespace (none found) and then look for it across all services in the system. We find one and there is no ambiguity (so far), and this service is from kubernetes. So we import it
- Same can be done for s2.ns2.
- Now, we try to look for google.com. None found in ns3 or istio-config. Then we look in all other namespaces. We find two google.coms (one in ns1 and one in ns2). Neither of these namespaces have been imported, and neither of these are from the Kubernetes service registry. At this point, rather than arbitrarily picking one variant, our options are ignore the host completely (the caller will get a 404 when traffic goes to google.com destination).

Reply:
@rshriram , wouldnt we pick randomly in the google case currently if you didn't have a sidecar object?

Reply:
I see your point. Thanks for clarifying. I am not sure what to do here then... The thing I don't like about the "pick it if it is not ambiguous" is that it can randomly break when someone adds a service entry. in this case, which google.com to use is completely arbitrary, so I think its fine to pick a random one? There is already similar cases for Services where we will pick a random one if its ambiguous. If the consumer wants to be explicit, then they can add `ns2/google.com` as an egress rule and declare which one they want.

Reply:
Ok, I'm going to make it more obvious and write a unit test, but it sounds like we're happy with random for now?

Reply:
Fine. So let us just pick something randomly.